mcpserver/server.go

```diff
@@ -0,0 +1,465 @@
+)
+
+// MattermostMCPServer provides a high-level interface for creating an MCP server
+// with Mattermost-specific tools and authentication
+type MattermostMCPServer struct {
+	mcpServer    *server.MCPServer
+	authProvider AuthenticationProvider
+	logger       mlog.LoggerIFace
+	config       Config
+}
+
+// NewMattermostMCPServer creates a new Mattermost MCP server with the specified configuration
+func NewMattermostMCPServer(config Config, authProvider AuthenticationProvider, logger mlog.LoggerIFace) (*MattermostMCPServer, error) {
+	// Create the mcp-go server directly
+	mcpServer := server.NewMCPServer(
+		"mattermost-mcp-server",
+		"0.1.0",
+		server.WithToolCapabilities(true), // Enable tool list changed notifications
+		server.WithLogging(),              // Enable logging capabilities
+	)
+
+	mattermostServer := &MattermostMCPServer{
+		mcpServer:    mcpServer,
+		authProvider: authProvider,
+		logger:       logger,
+		config:       config,
+	}
+
+	// For standalone mode (stdio with PAT), validate token at startup
+	if config.Transport == "stdio" {
+		ctx := context.Background()
+		_, err := authProvider.ValidateAuth(ctx, config.PersonalAccessToken)
+		if err != nil {
+			return nil, fmt.Errorf("startup token validation failed: %w", err)
+		}
+	}
+
+	// Register all Mattermost tools
+	mattermostServer.registerMattermostTools()
+
+	return mattermostServer, nil
+}
+
+// Serve starts the server using the configured transport
+func (s *MattermostMCPServer) Serve() error {
+	switch s.config.Transport {
+	case "stdio", "": // default to stdio for backward compatibility
+		return s.serveStdio()
+	case "http":
+		return s.serveHTTP()
+	default:
+		return fmt.Errorf("unsupported transport type: %s", s.config.Transport)
+	}
+}
+
+// serveStdio starts the server using stdio transport
+func (s *MattermostMCPServer) serveStdio() error {
+	// Configure error logger to use our mlog logger if available
+	var errorLogger *log.Logger
+	if s.logger != nil {
+		// Create a custom writer that forwards to mlog
+		errorLogger = log.New(&mlogWriter{logger: s.logger}, "", 0)
+	} else {
+		errorLogger = log.New(os.Stderr, "", log.LstdFlags)
+	}
+
+	return server.ServeStdio(s.mcpServer, server.WithErrorLogger(errorLogger))
+}
+
+// serveHTTP starts the server using HTTP transport
+func (s *MattermostMCPServer) serveHTTP() error {
+	// TODO: Implement HTTP/SSE transport for OAuth authentication
+	// This will be implemented when OAuth support is added
+	s.logger.Info("HTTP transport requested but not yet implemented")
+	s.logger.Info("Future implementation will support OAuth authentication and StreamableHTTP")
+	return fmt.Errorf("HTTP transport not yet implemented - will be added for OAuth support")
+}
+
+// registerMattermostTools registers all Mattermost tools with the MCP server
+func (s *MattermostMCPServer) registerMattermostTools() {
+	// Register read_post tool
+	readPostTool := mcp.NewTool("read_post",
```

Comment:
It might be nice if the registration and the implementation of the tool where in the same location. You can have a look at the mmtools package but it only solves the arguments part with struct tags.